src/umpire/resource/FileMemoryResource.cpp

```diff
@@ -0,0 +1,123 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+
+#include "umpire/resource/FileMemoryResource.hpp"
+
+#include <errno.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+
+namespace umpire {
+namespace resource {
+
+int FileMemoryResource::s_file_counter {0};
+
+FileMemoryResource::FileMemoryResource(
+    Platform platform, 
+    const std::string& name,
+    int id,
+    MemoryResourceTraits traits) :
+  MemoryResource{name, id, traits},
+  m_platform{platform},
+  m_size_map{}
+{
+} 
+
+FileMemoryResource::~FileMemoryResource()
+{
+  std::vector<void*> leaked_items;
+
+  for ( auto const& m : m_size_map ) {
+    leaked_items.push_back(m.first);
+  }
+
+  for ( auto const& p : leaked_items ) {
+    deallocate(p);
+  }
+}
+
+void* FileMemoryResource::allocate(std::size_t bytes)
+{
+  // Find output file directory for mmap files
+  const char* memory_file_dir{std::getenv("UMPIRE_MEMORY_FILE_DIR")};
+  std::string default_dir = "./";
+  if(memory_file_dir) {
+    default_dir = memory_file_dir;
+  }
+
+  // Create name and open file
+  std::stringstream ss;
+  ss << default_dir << "umpire_mem_" << getpid() << "_" << s_file_counter;
+  s_file_counter++;
+
+  int fd{open(ss.str().c_str(), O_RDWR | O_CREAT | O_LARGEFILE, S_IRWXU)};
+  if (fd == -1) { 
+    UMPIRE_ERROR("Opening File { " << ss.str() << " } Failed: " << strerror(errno)); 
+  }
+
+  // Setting Size Of Map File
+  const std::size_t pagesize{ (std::size_t) sysconf(_SC_PAGE_SIZE) };
+  std::size_t rounded_bytes{ ((bytes + (pagesize - 1))/ pagesize) * pagesize };
+
+  // Truncate file
+  int trun{ftruncate64(fd, rounded_bytes)};
+  if (trun == -1) { 
+    remove(ss.str().c_str()); 
```

Comment:
Part 1 of 2:
Since calling `remove` will cause the global variable `errno` to be modified we need to save it as follows:

```suggestion
    int errno_save = errno;
    remove(ss.str().c_str()); 
```